lib/rspec/support/differ.rb

```diff
@@ -13,12 +13,14 @@ class Differ
       def diff(actual, expected)
         diff = ""
 
+        no_procs_no_numbers = lambda {|var1, var2| no_procs?(var1, var2) && no_numbers?(var1, var2)}
+
         unless actual.nil? || expected.nil?
           if all_strings?(actual, expected)
-            if any_multiline_strings?(actual, expected)
-              diff = diff_as_string(coerce_to_string(actual), coerce_to_string(expected))
-            end
-          elsif no_procs?(actual, expected) && no_numbers?(actual, expected)
+            diff = diff_as_string(coerce_to_string(actual), coerce_to_string(expected)) if any_multiline_strings?(actual, expected)
+          elsif hash_with_anything?(expected)
```

Comment:
To reset the discussion below.
I am concerned about side effects of shared state in instance variables.
I’d live to get rid of an instance variable `@keys_with_anything`.
Only those two methods need it. Can we cache it somehow here in a regular variable? Can `hash_with_anything?` become something that returns a value that we can subsequently pass to `diff_as_object_with_anything`?